### PR TITLE
[Misc] Reduce `time vllm --help` and `time vllm serve --help` to <1s warm

### PR DESCRIPTION
**Pull request by netanel-haber**
This PR improves feedback time for CLI-only operations such as `--help` and
misspelled or misspecified arguments for top-level `vllm` and `vllm serve`.
It still builds the full generated parser.

**`vllm bench` performance remains out of scope.**

---

```shell
hyperfine --runs 10 --export-json "$RESULT" \
  -n cold "find . -type d -name __pycache__ -prune -exec rm -rf {} + && $COMMAND" \
  -n warm "$COMMAND"
```

Median of 10 runs on my RTX 3500 Ada laptop:

| command | mode | main (`df344bf36f`) | PR (`176a9b0527`) | speedup |
| --- | --- | ---: | ---: | ---: |
| `vllm serve --help` | cold | 23.72 s | 1.64 s | 14.4x |
| `vllm serve --help` | warm | 10.01 s | 0.755 s | 13.3x |
| `vllm --help` | cold | 23.37 s | 0.475 s | 49.2x |
| `vllm --help` | warm | 10.36 s | 0.164 s | 63.2x |

The regression test warms once, measures seven runs per command, and requires a
median below 1 second.

Help output is unchanged.

I verified that help strings are identical with [dump_help_strings.sh](https://github.com/user-attachments/files/27542342/dump_help_strings.sh):
```shell
git switch main 
dump_help_strings.sh /tmp/vllm-help.main 
git switch feature/cli-response-time 
dump_help_strings.sh /tmp/vllm-help.branch 
diff -u /tmp/vllm-help.main /tmp/vllm-help.branch
```

```shell
pytest tests/entrypoints/test_cli_response_times.py
pytest tests/config/test_multimodal_config.py -k mm_processor_device
pytest tests/test_zen_cpu_platform_detection.py
```

Codex assisted with conflict resolution, profiling, and validation.

### CODE REVIEW

**Review by aarnphm — COMMENTED** [review]
I'm a bit hesitant to lazy load imports like this, given that we will end-up playing a game of whackamole and people aren't aware of these.

https://peps.python.org/pep-0810/ so i would prefer to just waiting for 3.15, then we can to proper lazy import here. 

also will ask other commiters for their opinion as well

**Comment by aarnphm** [comment]
Another thing is that we will have to load/import these anyway, so I'm fine waiting 5s in the beginning to start the server

**Comment by netanel-haber** [comment]
> Another thing is that we will have to load/import these anyway, so I'm fine waiting 5s in the beginning to start the server

The goal of this PR is narrower: it doesn’t eliminate work, it just avoids doing that work before we know the user actually needs it. That makes pure CLI feedback much faster.
I’ve personally run `--help` or passed a misspelled parameter and waited ~20s, not exaggerating, just to get that feedback. Others have mentioned to me saving the help output to a file so they don’t have to keep waiting for it. So from my perspective, this is mainly a QoL improvement for users.

That said, I understand if the maintainers feel the extra complexity isn’t worth it. I’ll defer to you and the rest of the team on that

**Comment by netanel-haber** [comment]
> https://peps.python.org/pep-0810/ so i would prefer to just waiting for 3.15, then we can to proper lazy import here. 

The issue is that Python3.11+, and 3.13+ without a doubt, will be supported by vllm for long after 3.15 is stable

**Comment by netanel-haber** [comment]
> I'm a bit hesitant to lazy load imports like this, given that we will end-up playing a game of whackamole and people aren't aware of these.

That is where the test I added comes in, it should make it fool proof to know when you've mistakenly introduced `import torch` into the cli hot path

**Comment by netanel-haber** [comment]
In addition, even with lazy imports, people can still put torch.dtype etc. in the top level and trigger an import, and this will prevent that from happening

**Review by NickLucche — COMMENTED** [review]
cc @DarkLight1337 @hmellor

File: vllm/entrypoints/cli/main.py [hunk 1]
```diff
@@ -33,7 +34,9 @@ def main():
         vllm.entrypoints.cli.run_batch,
     ]
 
-    cli_env_setup()
+    if "VLLM_WORKER_MULTIPROC_METHOD" not in os.environ:
+        logger.debug("Setting VLLM_WORKER_MULTIPROC_METHOD to 'spawn'")
+        os.environ["VLLM_WORKER_MULTIPROC_METHOD"] = "spawn"
```
Comment (hmellor) [inline vllm/entrypoints/cli/main.py]: What if `cli_env_setup` changes in future? We would need to know to duplicate that change here?

File: vllm/entrypoints/cli/openai.py [hunk 2]
```diff
@@ -1,18 +1,20 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
+from __future__ import annotations
+
```
Comment (hmellor) [inline vllm/entrypoints/cli/openai.py]: Can we just use `"OpenAI"` instead?

File: vllm/entrypoints/cli/launch.py [hunk 3]
```diff
@@ -84,6 +80,9 @@ def subparser_init(
             description=DESCRIPTION,
             usage=f"vllm {self.name} <component> [options]",
         )
+        if not is_cli_subcommand(self.name):
+            return launch_parser
```
Comment (hmellor) [inline vllm/entrypoints/cli/launch.py]: Instead of adding this boilerplate to every `subparser_init` method could we just filter the calls before we enter these methods?

i.e. here:

https://github.com/vllm-project/vllm/blob/85959567c3e71a9965616ebebe1853ca48d8d20f/vllm/entrypoints/cli/main.py#L85

File: vllm/engine/arg_utils.py [hunk 4]
```diff
@@ -2464,7 +2506,10 @@ def add_cli_args(
             "- DEBUG: Prompt inputs (e.g: text, token IDs).\n"
             "You can set the minimum log level via `VLLM_LOGGING_LEVEL`.",
         )
-        current_platform.pre_register_and_update(parser)
+        if load_plugins_by_group(PLATFORM_PLUGINS_GROUP):
+            from vllm.platforms import current_platform
+
+            current_platform.pre_register_and_update(parser)
```
[the remark this answers is not in the document]
Comment (netanel-haber) [reply vllm/engine/arg_utils.py:2958]: Built-in platforms do not override the `pre_register_and_update` hook, so there is no need to load current_platform for them and pay the cost.

File: vllm/entrypoints/cli/launch.py (hunk 3 above)
Comment (netanel-haber) [reply vllm/entrypoints/cli/launch.py]: Done

File: vllm/entrypoints/cli/main.py (hunk 1 above)
Comment (netanel-haber) [reply vllm/entrypoints/cli/main.py]: I reinstated cli_env_setup, moved it to `vllm/entrypoints/cli/__init__.py` so it doesn't have such a heavy import cost.

**Comment by netanel-haber** [comment]
Ready for review again.

**Review by NickLucche — COMMENTED** [review]
I only managed to take a look now @netanel-haber , thanks for the work.
I think I was expecting a much bigger radius, I would actually like to give CI a spin after rebase.
The only thing that concerns me though is how do we keep the lazy imports in place overtime, given we might easily regress to current state with a few misplaced imports.

**Comment by matteso1** [comment]
Hi @netanel-haber, I have a narrower CLI-dispatch patch ready that overlaps the user-facing goal here, so I wanted to coordinate before opening it.

I based the shape on two review points from @hmellor in this PR: [filter once in `cli/main.py`](https://github.com/vllm-project/vllm/pull/41518#discussion_r3271828639) instead of adding boilerplate to every subparser, and [keep `cli_env_setup` shared](https://github.com/vllm-project/vllm/pull/41518#discussion_r3271797197). The boundary is different. #41518 changes 39 files to make the config and platform import graph light enough to preserve byte-identical full help, and it leaves bench for a follow-up. My patch changes 14 files, loads only the selected command from one lightweight registry, includes bench, and makes plain `vllm serve --help` a compact core view. `vllm serve --help=all` keeps the canonical parser and complete option list.

Against `76fb6d210a`, fresh-process medians on an Apple M5 Pro after one unrecorded warm-up per arm and command were 3094.027 ms versus 44.967 ms for `vllm --help`, and 3089.799 ms versus 44.644 ms for plain serve help, n=5 per arm. The focused CLI/CUDA suite passed 45 tests, and the parser/runtime-override suites passed 56.

I plan to open that narrower patch as the generic CLI prerequisite Simon asked me to split out of #51360, and I will cite this PR and its review history in the body. The full-help/config-annotation design here remains a separate tradeoff because my patch deliberately changes plain serve help. If that boundary cuts across where you want to take #41518, I am happy to reconcile the overlap in the PR.

**Comment by netanel-haber** [comment]
> Hi @netanel-haber, I have a narrower CLI-dispatch patch ready that overlaps the user-facing goal here, so I wanted to coordinate before opening it.

Hey, I think you should push whatever you think is best, but also notice that my pr not only reduces `--help` to `<1s`, it also reduces time for misspelled, or arguments whose values are the wrong type etc:

> This pr improves the feedback time of pure CLI operations (`--help`, Misspelled flags, flags that don't exist, etc.) for top-level vllm and vllm serve.
